client/writer.go

```diff
@@ -0,0 +1,376 @@
+		w.Unlock()
+		return errInstanceWriterClosed
+	}
+	encoder, exists = w.encodersByShard[shard]
+	if exists {
+		err := w.encodeUntimedWithLock(encoder, metric, metadatas)
+		w.Unlock()
+		return err
+	}
+	encoder = w.newLockedEncoderFn(w.encoderOpts)
+	w.encodersByShard[shard] = encoder
+	err := w.encodeUntimedWithLock(encoder, metric, metadatas)
+	w.Unlock()
+
+	return err
+}
+
+func (w *writer) Flush() error {
+	w.RLock()
+	if w.closed {
+		w.RUnlock()
+		return errInstanceWriterClosed
+	}
+	err := w.flushWithLock()
+	w.RUnlock()
+
+	if err != nil {
+		w.metrics.flushErrors.Inc(1)
+		return err
+	}
+	return nil
+}
+
+func (w *writer) Close() error {
+	w.Lock()
+	defer w.Unlock()
+
+	if w.closed {
+		return errInstanceWriterClosed
+	}
+	w.closed = true
+	if err := w.flushWithLock(); err != nil {
+		w.metrics.flushErrors.Inc(1)
+	}
+	return w.queue.Close()
+}
+
+func (w *writer) encodeUntimedWithLock(
+	encoder *lockedEncoder,
+	metric unaggregated.MetricUnion,
+	metadatas metadata.StagedMetadatas,
+) error {
+	encoder.Lock()
+
+	var (
+		sizeBefore = encoder.Len()
+		encodeErr  error
+		enqueueErr error
+	)
+	switch metric.Type {
+	case unaggregated.CounterType:
+		msg := encoding.UnaggregatedMessageUnion{
+			Type: encoding.CounterWithMetadatasType,
+			CounterWithMetadatas: unaggregated.CounterWithMetadatas{
+				Counter:         metric.Counter(),
+				StagedMetadatas: metadatas,
+			}}
+		encodeErr = encoder.EncodeMessage(msg)
+	case unaggregated.BatchTimerType:
+		// If there is no limit on the timer batch size, write the full batch.
+		if w.maxTimerBatchSize == 0 {
+			msg := encoding.UnaggregatedMessageUnion{
+				Type: encoding.BatchTimerWithMetadatasType,
+				BatchTimerWithMetadatas: unaggregated.BatchTimerWithMetadatas{
+					BatchTimer:      metric.BatchTimer(),
+					StagedMetadatas: metadatas,
+				}}
+			encodeErr = encoder.EncodeMessage(msg)
+			break
+		}
+
+		// Otherwise, honor maximum timer batch size.
+		var (
+			batchTimer     = metric.BatchTimer()
+			timerValues    = batchTimer.Values
+			numTimerValues = len(timerValues)
+			start, end     int
+		)
+		for start = 0; start < numTimerValues; start = end {
+			end = start + w.maxTimerBatchSize
+			if end > numTimerValues {
+				end = numTimerValues
+			}
+			singleBatchTimer := unaggregated.BatchTimer{
+				ID:     batchTimer.ID,
+				Values: timerValues[start:end],
+			}
+			msg := encoding.UnaggregatedMessageUnion{
+				Type: encoding.BatchTimerWithMetadatasType,
+				BatchTimerWithMetadatas: unaggregated.BatchTimerWithMetadatas{
+					BatchTimer:      singleBatchTimer,
+					StagedMetadatas: metadatas,
+				}}
+			encodeErr = encoder.EncodeMessage(msg)
+			if encodeErr != nil {
+				break
+			}
+
+			// If the buffer isn't big enough continue to the next iteration.
+			if sizeAfter := encoder.Len(); sizeAfter < w.flushSize {
+				continue
+			}
+
+			// Otherwise we enqueue the current buffer.
+			buffer := w.prepareEnqueueBufferWithLock(encoder, sizeBefore)
```

Comment:
why is this function called for timers specifically than reuse the logic outside the loop?

Reply:
right this is still in the loop, make sense